src/statecharts/impl.cljc

```diff
@@ -816,6 +817,7 @@
          event {:type :fsm/init}
          [_state actions] (-do-init fsm)
          state (assoc context
+                 :_id id
```

Comment:
Let's discuss this a bit more... I'm trying to keep this PR as narrow as possible, but there is at least [one other](https://github.com/lucywang000/clj-statecharts/blob/4530c41bd72b9307241ae3abab13899ace54a776/src/statecharts/service.cljc#L19-L20) place in the code that couples a machine to a single state. If you ever wanted to provide a Service where one machine managed several states, you'd run into this same problem again. Without something like `:_id` on individual states, all schedulers couple a machine to a single state. So, fixing that in the main implementation of IScheduler is a start to fixing it elsewhere.

It's not _wrong_ that Service and re-frame couple a machine to a single state, but it is limiting. When I first read the `clj-statecharts` [documentation](https://lucywang000.github.io/clj-statecharts/docs/concepts/#machine--state) it was a major insight that state and machine can and should be separate. It's a beautiful way to decouple the problem and I think statecharts should highlight and encourage that perspective.

As far as `:_id` vs `:_rf-path` goes, I was thinking the converse. If all states had an `:_id`, then we wouldn't need `:_rf-path`. We could use the `:_id` in the app-db path.

And as far as `:_id` being dead weight... yes, it adds 16 bytes to every state (or less if a user provides something other than a uuid). But it's part of ensuring you don't have to store a copy of the _entire_ state machine alongside every state. So in terms of size, it's a win overall. And I appreciate wanting to hide data from users, but the docs make it clear that the underscored variables are reserved and shouldn't be messed with. And though a user could read the `:_id`, there's not much they could _do_ with it.

So, I'm going to leave `:_id` and the timeout cache as is for now. Ultimately it's up to you and I'm happy to change it if you want. Let me know.

Reply:
Thanks for the insights, but still I'd prefer not do add this `_id` field unconditionally, for these concerns:

1) by definition a state is just an value, and it should not have to have anything to do with a "unique id".

2) the unique id only make sense when used in a scheduler, e.g. managing long-running operations like http requests, but those are not all the use cases for a state machine library. For instance I have quite a few projects that only uses the functional layer of this library, and I would not like a uuid being inserted to my state map in those cases, because it would be pure noise when I inspect the data. (the latest one of such use cases is some sample code I just write for fun [here](https://github.com/noprompt/meander/discussions/216#discussioncomment-1563336).

> but the docs make it clear that the underscored variables are reserved and shouldn't be messed with.

The doc suggests the user could read and make use of these reserved keys, like `_state`, `_prev-state`, just not to update them.

Reply:
Fair enough. I'll take the _id out. 

Without it, the new `statecharts.impl-test/test-simultaneous-delays` test fails. The [last assertion](https://github.com/lucywang000/clj-statecharts/pull/8/files#diff-c21ec469c0177ad9e6647bdc64da038db26a91c9372fc7b4c38ab6649b21ab9eR1047) breaks because we're back to the situation where the timeout id cache gets clobbered. I can think of several ways to address that:
1. Delete the test. Though, without it, the other changes in this PR feel pretty un-motivated.
2. Re-implement IScheduler in the test. Not my favorite. I don't like tests that dig deep into the implementation. They require a lot of knowledge to maintain and hinder refactoring.
3. Re-implement IScheduler in `statecharts.delayed`. The main Scheduler wouldn't need _id, but the new implementation would. (Although I'd probably make _id configurable, so users could specify _id, _rf-path, or whatever suits them.)

Number 3 is my preference for a few reasons. First, it would be more likely to stay in sync with changes to the main Scheduler. Second, I could take advantage of it in my re-frame integration. And third, it would be useful for anyone else who runs into problems managing multiple states.

Would you be OK with the third option?

Oh, and also, that parser code is pretty cool! Very nice!

Reply:
I started to implement 3, but ended up with a 4th option, which you can see in 3ecd2f2. Instead of creating a new IScheduler implementation, it adapts the existing Scheduler to be able to manage many states, if so configured. If you'd like, I can split it into two implementations, but I think I prefer having a single code path for the scheduler internals.

Reply:
Yeah the `context-id` function is a very clever design :P